### PR TITLE
To fix the ios config exception where it was being ignored and play run was successful

### LINKED ISSUE #213: Command failed to run but reported success
##### SUMMARY
I am testing static route addition, On the Cisco router but the command fails on the router but Ansible reports that 1 Change is completed. It is not catching and reporting the errors produced by IOS CLI.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
cisco.ios.ios_static_routes

##### ANSIBLE VERSION

```ansible 2.9.6
  config file = /home/ravinder/automate/ansible.cfg
  configured module search path = ['/home/ravinder/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
```

##### CONFIGURATION

```paste below
DEFAULT_TIMEOUT(/home/ravinder/automate/ansible.cfg) = 5
DEFAULT_TRANSPORT(/home/ravinder/automate/ansible.cfg) = paramiko
HOST_KEY_CHECKING(/home/ravinder/automate/ansible.cfg) = False
```

##### OS / ENVIRONMENT

Cisco IOS Version 15.5(2)T

##### STEPS TO REPRODUCE

Either use a wrong VRF management name
Or use Next-hop as router self IP
I am sure there will be other cases where the command fails for some reason by the router but Ansible is not reporting it.
In short, it is not catching and reporting the errors produced by IOS CLI.
It will fail on the router but Ansible will report success.
 

```yaml
- name: Merge provided configuration with device configuration
          cisco.ios.ios_static_routes:
            config:
            - vrf: Management
              address_families:
              - afi: ipv4
                routes:
                - dest: 192.0.2.0/24
                  next_hops:
                  - forward_router_address: 10.1.255.6
                    name: Test_Static_Route
                    tag: 50
            - address_families:
              - afi: ipv4
                routes:
                - dest: 198.51.100.0/24
                  next_hops:
                  - forward_router_address: 10.1.255.13
                    name: merged_route_1
                    distance_metric: 110
                    tag: 40
                    multicast: true
                  - forward_router_address: 10.1.255.13
                    name: merged_route_2
                    distance_metric: 30
                  - forward_router_address: 10.1.255.13
                    name: merged_route_3
            state: merged
```

##### EXPECTED RESULTS

Expecting an error thrown by Cisco IOS reported back.

##### ACTUAL RESULTS

```paste below

PLAY [Network Route Change Playbook] ********************************************************************************

TASK [Gathering Facts] **********************************************************************************************
ok: [R1]

TASK [OBTAIN LOGIN INFORMATION] *************************************************************************************
ok: [R1]

TASK [Create change dir] ********************************************************************************************
ok: [R1]

TASK [Merge provided configuration with device configuration] *******************************************************
changed: [R1]

TASK [SAVE & BACKUP CONFIGURATION] **********************************************************************************
ok: [R1]

PLAY RECAP **********************************************************************************************************
R1                         : ok=5    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0


```

### PR DESCRIPTION
**Pull request by justjais**
##### SUMMARY

To fix the ios config exception where it was being ignored and play run was successful. fixes #213, #258 n #81  

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ios_*

##### ADDITIONAL INFORMATION

With the changes now the play run will halt and a warning/error will be thrown instead of giving a false impression to the user that the play run was successful and config is applied to the box.

```paste below
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.module_utils.connection.ConnectionError: Command: 'address-family ipv4 multicast vrf blue' results into: '% VRF blue does not exist.'

[WARNING]: Platform darwin on host an-ios-01.ansible.eng.rdu2.redhat.com is using the discovered Python interpreter at /usr/bin/python, but future installation of another Python interpreter could change the
meaning of that path. See https://docs.ansible.com/ansible/2.10/reference_appendices/interpreter_discovery.html for more information.
fatal: [an-ios-01.ansible.eng.rdu2.redhat.com]: FAILED! => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python"}, "changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/Users/sjaiswal/.ansible/tmp/ansible-local-2688rYazvn/ansible-tmp-1614329642.43-2744-122064453824984/AnsiballZ_ios_config.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/Users/sjaiswal/.ansible/tmp/ansible-local-2688rYazvn/ansible-tmp-1614329642.43-2744-122064453824984/AnsiballZ_ios_config.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/Users/sjaiswal/.ansible/tmp/ansible-local-2688rYazvn/ansible-tmp-1614329642.43-2744-122064453824984/AnsiballZ_ios_config.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_config', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py\", line 188, in run_module\n    fname, loader, pkg_name)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py\", line 82, in _run_module_code\n    mod_name, mod_fname, mod_loader, pkg_name)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py\", line 72, in _run_code\n    exec code in run_globals\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_KSgyed/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py\", line 593, in <module>\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_KSgyed/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py\", line 518, in main\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_KSgyed/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py\", line 385, in edit_config_or_macro\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_KSgyed/ansible_cisco.ios.ios_config_payload.zip/ansible/module_utils/connection.py\", line 195, in __rpc__\nansible.module_utils.connection.ConnectionError: Command: 'address-family ipv4 multicast vrf blue' results into: '% VRF blue does not exist.'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```
or, `with -vvvv`

```
The full traceback is:
Traceback (most recent call last):
  File "/Users/sjaiswal/.ansible/tmp/ansible-local-2861LxRqOz/ansible-tmp-1614329737.56-2917-87656613359497/AnsiballZ_ios_config.py", line 102, in <module>
    _ansiballz_main()
  File "/Users/sjaiswal/.ansible/tmp/ansible-local-2861LxRqOz/ansible-tmp-1614329737.56-2917-87656613359497/AnsiballZ_ios_config.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/Users/sjaiswal/.ansible/tmp/ansible-local-2861LxRqOz/ansible-tmp-1614329737.56-2917-87656613359497/AnsiballZ_ios_config.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_config', init_globals=None, run_name='__main__', alter_sys=True)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 188, in run_module
    fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py", line 593, in <module>
  File "/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py", line 518, in main
  File "/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py", line 385, in edit_config_or_macro
  File "/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible/module_utils/connection.py", line 195, in __rpc__
ansible.module_utils.connection.ConnectionError: Command: 'address-family ipv4 multicast vrf blue' results into: '% VRF blue does not exist.'
[WARNING]: Platform darwin on host an-ios-01.ansible.eng.rdu2.redhat.com is using the discovered Python interpreter at /usr/bin/python, but future installation of another Python interpreter could change the
meaning of that path. See https://docs.ansible.com/ansible/2.10/reference_appendices/interpreter_discovery.html for more information.
fatal: [an-ios-01.ansible.eng.rdu2.redhat.com]: FAILED! => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/Users/sjaiswal/.ansible/tmp/ansible-local-2861LxRqOz/ansible-tmp-1614329737.56-2917-87656613359497/AnsiballZ_ios_config.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/Users/sjaiswal/.ansible/tmp/ansible-local-2861LxRqOz/ansible-tmp-1614329737.56-2917-87656613359497/AnsiballZ_ios_config.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/Users/sjaiswal/.ansible/tmp/ansible-local-2861LxRqOz/ansible-tmp-1614329737.56-2917-87656613359497/AnsiballZ_ios_config.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_config', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py\", line 188, in run_module\n    fname, loader, pkg_name)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py\", line 82, in _run_module_code\n    mod_name, mod_fname, mod_loader, pkg_name)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py\", line 72, in _run_code\n    exec code in run_globals\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py\", line 593, in <module>\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py\", line 518, in main\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_config.py\", line 385, in edit_config_or_macro\n  File \"/var/folders/n7/73s22vps0ls44ng__t8mwtl40000gn/T/ansible_cisco.ios.ios_config_payload_qXykBM/ansible_cisco.ios.ios_config_payload.zip/ansible/module_utils/connection.py\", line 195, in __rpc__\nansible.module_utils.connection.ConnectionError: Command: 'address-family ipv4 multicast vrf blue' results into: '% VRF blue does not exist.'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

```

### CODE REVIEW

File: plugins/cliconf/ios.py [hunk 1]
```diff
@@ -198,6 +198,16 @@ def edit_config(
 
                 cmd = line["command"]
                 if cmd != "end" and cmd[0] != "!":
+                    if (
+                        self.send_command(**line)
+                        and len(self.send_command(**line).split(" ")) > 1
+                        or "%" in self.send_command(**line).split(" ")
```
Comment (ganeshrn) [inline plugins/cliconf/ios.py:204]: This behaviour change seems backwards-incompatible for `ios_config`/`cli_config` modules as warning from the device is now treated as an error. Can this scenario be handled in the respective resource modules?

File: plugins/cliconf/ios.py [hunk 2]
```diff
@@ -198,6 +198,16 @@ def edit_config(
 
                 cmd = line["command"]
```
Comment (NilashishC) [inline plugins/cliconf/ios.py:199]: I believe an alternate and cleaner fix would be to update the terminal error regex.

**Comment by justjais** [comment]
recheck

**Comment by Andersson007** [comment]
```
2021-03-26 14:46:30.964016 | ubuntu-bionic | The full traceback is:
2021-03-26 14:46:30.964069 | ubuntu-bionic | Traceback (most recent call last):
2021-03-26 14:46:30.964080 | ubuntu-bionic |   File "/home/zuul/.ansible/tmp/ansible-local-107266GH68E/ansible-tmp-1616769987.14-11181-75249604739532/AnsiballZ_ios_bgp_global.py", line 116, in <module>
2021-03-26 14:46:30.964089 | ubuntu-bionic |     _ansiballz_main()
2021-03-26 14:46:30.964097 | ubuntu-bionic |   File "/home/zuul/.ansible/tmp/ansible-local-107266GH68E/ansible-tmp-1616769987.14-11181-75249604739532/AnsiballZ_ios_bgp_global.py", line 108, in _ansiballz_main
2021-03-26 14:46:30.964106 | ubuntu-bionic |     invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
2021-03-26 14:46:30.964114 | ubuntu-bionic |   File "/home/zuul/.ansible/tmp/ansible-local-107266GH68E/ansible-tmp-1616769987.14-11181-75249604739532/AnsiballZ_ios_bgp_global.py", line 54, in invoke_module
2021-03-26 14:46:30.964140 | ubuntu-bionic |     runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_bgp_global', init_globals=None, run_name='__main__', alter_sys=True)
2021-03-26 14:46:30.964148 | ubuntu-bionic |   File "/usr/lib/python2.7/runpy.py", line 188, in run_module
2021-03-26 14:46:30.964174 | ubuntu-bionic |     fname, loader, pkg_name)
2021-03-26 14:46:30.964183 | ubuntu-bionic |   File "/usr/lib/python2.7/runpy.py", line 82, in _run_module_code
2021-03-26 14:46:30.964191 | ubuntu-bionic |     mod_name, mod_fname, mod_loader, pkg_name)
2021-03-26 14:46:30.964200 | ubuntu-bionic |   File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
2021-03-26 14:46:30.964208 | ubuntu-bionic |     exec code in run_globals
2021-03-26 14:46:30.964216 | ubuntu-bionic |   File "/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_global.py", line 2138, in <module>
2021-03-26 14:46:30.964225 | ubuntu-bionic |   File "/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_global.py", line 2133, in main
2021-03-26 14:46:30.964234 | ubuntu-bionic |   File "/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/config/bgp_global/bgp_global.py", line 68, in execute_module
2021-03-26 14:46:30.964244 | ubuntu-bionic |   File "/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/rm_base/resource_module.py", line 181, in run_commands
2021-03-26 14:46:30.964253 | ubuntu-bionic |   File "/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible/module_utils/connection.py", line 195, in __rpc__
2021-03-26 14:46:30.964267 | ubuntu-bionic | ansible.module_utils.connection.ConnectionError: no neighbor 198.51.100.1 description merge neighbor
2021-03-26 14:46:30.964276 | ubuntu-bionic | % Specify remote-as or peer-group commands first
2021-03-26 14:46:30.964284 | ubuntu-bionic | R1(config-router)#
2021-03-26 14:46:30.965086 | ubuntu-bionic | fatal: [ios-15.6-2T]: FAILED! => {
2021-03-26 14:46:30.965111 | ubuntu-bionic |     "changed": false,
2021-03-26 14:46:30.965123 | ubuntu-bionic |     "module_stderr": "Traceback (most recent call last):\n  File \"/home/zuul/.ansible/tmp/ansible-local-107266GH68E/ansible-tmp-1616769987.14-11181-75249604739532/AnsiballZ_ios_bgp_global.py\", line 116, in <module>\n    _ansiballz_main()\n  File \"/home/zuul/.ansible/tmp/ansible-local-107266GH68E/ansible-tmp-1616769987.14-11181-75249604739532/AnsiballZ_ios_bgp_global.py\", line 108, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/zuul/.ansible/tmp/ansible-local-107266GH68E/ansible-tmp-1616769987.14-11181-75249604739532/AnsiballZ_ios_bgp_global.py\", line 54, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_bgp_global', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python2.7/runpy.py\", line 188, in run_module\n    fname, loader, pkg_name)\n  File \"/usr/lib/python2.7/runpy.py\", line 82, in _run_module_code\n    mod_name, mod_fname, mod_loader, pkg_name)\n  File \"/usr/lib/python2.7/runpy.py\", line 72, in _run_code\n    exec code in run_globals\n  File \"/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_global.py\", line 2138, in <module>\n  File \"/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_global.py\", line 2133, in main\n  File \"/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/config/bgp_global/bgp_global.py\", line 68, in execute_module\n  File \"/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/rm_base/resource_module.py\", line 181, in run_commands\n  File \"/tmp/ansible_cisco.ios.ios_bgp_global_payload_Drsdge/ansible_cisco.ios.ios_bgp_global_payload.zip/ansible/module_utils/connection.py\", line 195, in __rpc__\nansible.module_utils.connection.ConnectionError: no neighbor 198.51.100.1 description merge neighbor\r\n% Specify remote-as or peer-group commands first\r\nR1(config-router)#\n",
2021-03-26 14:46:30.965141 | ubuntu-bionic |     "module_stdout": "",
2021-03-26 14:46:30.965150 | ubuntu-bionic |     "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
2021-03-26 14:46:30.965158 | ubuntu-bionic |     "rc": 1
2021-03-26 14:46:30.965166 | ubuntu-bionic | }
```
@justjais is it possible to fix this?

**Comment by Andersson007** [comment]
recheck

**Comment by justjais** [comment]
recheck

**Comment by KB-perByte** [comment]
recheck

**Pull request closed without merging**